File: agario/bots/manager.py
"""Bot plugin manager and runtime orchestration."""

from __future__ import annotations

import logging
import random
from collections import defaultdict
from dataclasses import dataclass
from typing import TYPE_CHECKING

from .. import config
from .registry import BotRegistry, load_plugin_modules
from .types import (
    BlobView,
    BotAction,
    BotContext,
    BotInitContext,
    BotSpec,
    EjectedView,
    FoodView,
    PlayerView,
    VirusView,
)

if TYPE_CHECKING:
    from ..world import GameWorld

logger = logging.getLogger(__name__)


def parse_bot_specs(raw: str) -> list[BotSpec]:
    specs: list[BotSpec] = []
    cleaned = (raw or "").strip()
    if not cleaned:
        return specs

    for chunk in cleaned.split(","):
        part = chunk.strip()
        if not part:
            continue
        tokens = [t.strip() for t in part.split(":")]
        if not tokens[0]:
            raise ValueError(f"Invalid bot spec '{part}': plugin name is required")

        plugin_name = tokens[0].lower()
        count = 1
        team_id: str | None = None
        name_prefix: str | None = None

        if len(tokens) >= 2 and tokens[1]:
            count = int(tokens[1])
        if len(tokens) >= 3 and tokens[2] and tokens[2] != "-":
            team_id = tokens[2]
        if len(tokens) >= 4 and tokens[3] and tokens[3] != "-":
            name_prefix = tokens[3]

        if count <= 0:
            raise ValueError(f"Invalid bot spec '{part}': count must be > 0")

        specs.append(BotSpec(plugin_name=plugin_name, count=count, team_id=team_id, name_prefix=name_prefix))

    return specs


@dataclass(slots=True)
class _BotAgent:
    player_id: str
    plugin_name: str
    team_id: str | None
    name_prefix: str
    brain: object
    memory: dict


class BotManager:
    def __init__(
        self,
        world: GameWorld,
        *,
        enabled: bool,
        plugin_modules: tuple[str, ...],
        bot_specs: list[BotSpec],
        seed: int,
    ) -> None:
        self.world = world
        self.enabled = enabled
        self.plugin_modules = plugin_modules
        self.bot_specs = bot_specs
        self._started = False
        self._registry = BotRegistry()
        self._rng = random.Random(seed)
        self._agents: dict[str, _BotAgent] = {}
        self._team_state: dict[str, dict] = {}
        self._spawn_index = 0

    @classmethod
    def from_config(cls, world: GameWorld) -> BotManager:
        specs = parse_bot_specs(config.BOT_SPECS)
        return cls(
            world,
            enabled=config.BOTS_ENABLED,
            plugin_modules=config.BOT_PLUGIN_MODULES,
            bot_specs=specs,
            seed=config.BOT_RANDOM_SEED,
        )

    def describe(self) -> dict:
        return {
            "enabled": self.enabled,
            "started": self._started,
            "pluginModules": list(self.plugin_modules),
            "registeredPlugins": list(self._registry.names),
            "botSpecs": [
                {
                    "plugin": spec.plugin_name,
                    "count": spec.count,
                    "team": spec.team_id,
                    "namePrefix": spec.name_prefix,
                }
                for spec in self.bot_specs
            ],
            "activeBots": len(self._agents),
            "spawnOnEaten": config.BOT_SPAWN_ON_EATEN,
            "spawnPerElimination": config.BOT_SPAWN_PER_ELIMINATION,
            "maxActiveBots": config.BOT_MAX_ACTIVE,
        }

    def ensure_started(self, now: float) -> None:
        if self._started or not self.enabled:
            return

        load_plugin_modules(self.plugin_modules, self._registry)
        for spec in self.bot_specs:
            for i in range(spec.count):
                name_prefix = spec.name_prefix or spec.plugin_name.replace("_", " ").title()
                player_name = self._build_name(spec, i)
                self._spawn_bot(
                    plugin_name=spec.plugin_name,
                    team_id=spec.team_id,
                    name_prefix=name_prefix,
                    now=now,
                    explicit_name=player_name,
                )

        self._started = True
        logger.info("BotManager started with %d bots using plugins: %s", len(self._agents), ", ".join(self._registry.names))

    def tick(self, dt: float, now: float) -> None:
        if not self._started or not self._agents:
            return

        players_by_id, players, foods, ejected, viruses = self._build_views()
        if not players_by_id:
            return

        team_members: dict[str, list[str]] = defaultdict(list)
        for agent in self._agents.values():
            team_key = self._team_key(agent.plugin_name, agent.team_id)
            team_members[team_key].append(agent.player_id)
        for key, members in team_members.items():
            state = self._team_state.setdefault(key, {})
            state["members"] = tuple(sorted(pid for pid in members if pid in players_by_id))

        for player_id, agent in list(self._agents.items()):
            me = players_by_id.get(player_id)
            if me is None:
                # If a bot was explicitly removed, drop it from the bot manager.
                self._agents.pop(player_id, None)
                continue

            was_alive = bool(agent.memory.get("_was_alive", bool(me.blobs)))
            is_alive = bool(me.blobs)
            if not is_alive and was_alive:
                self._spawn_extra_on_elimination(agent, now)
            agent.memory["_was_alive"] = is_alive

            team_key = self._team_key(agent.plugin_name, agent.team_id)
            team_state = self._team_state.setdefault(team_key, {})
            ctx = BotContext(
                now=now,
                dt=dt,
                world_width=config.WORLD_WIDTH,
                world_height=config.WORLD_HEIGHT,
                me=me,
                players=players,
                foods=foods,
                ejected=ejected,
                viruses=viruses,
                team_state=team_state,
                memory=agent.memory,
            )

            try:
                action = agent.brain.decide(ctx)
            except Exception:
                logger.exception("Bot plugin '%s' failed for player %s", agent.plugin_name, player_id)
                action = self._fallback_action(me)

            self.world.set_input(
                player_id=player_id,
                target_x=self._clamp(action.target_x, 0.0, config.WORLD_WIDTH),
                target_y=self._clamp(action.target_y, 0.0, config.WORLD_HEIGHT),
                split=bool(action.split),
                eject=bool(action.eject),
            )

    def _build_views(
        self,
    ) -> tuple[
        dict[str, PlayerView],
        tuple[PlayerView, ...],
        tuple[FoodView, ...],
        tuple[EjectedView, ...],
        tuple[VirusView, ...],
    ]:
        players: list[PlayerView] = []
        for player in self.world.players.values():
            blobs = tuple(
                BlobView(
                    id=blob.id,
                    player_id=blob.player_id,
                    x=blob.x,
                    y=blob.y,
                    mass=blob.mass,
                    radius=blob.radius,
                )
                for blob in player.blobs.values()
            )
            players.append(
                PlayerView(
                    id=player.id,
                    name=player.name,
                    color=player.color,
                    is_bot=player.is_bot,
                    plugin_name=player.bot_plugin,
                    team_id=player.bot_team,
                    total_mass=sum(b.mass for b in blobs),
                    blobs=blobs,
                )
            )

        players.sort(key=lambda p: p.id)
        players_by_id = {p.id: p for p in players}

        foods = tuple(
            FoodView(
                id=food.id,
                x=food.x,
                y=food.y,
                mass=food.mass,
                radius=food.radius,
                color=food.color,
            )
            for food in self.world.foods.values()
        )
        ejected = tuple(
            EjectedView(
                id=item.id,
                x=item.x,
                y=item.y,
                mass=item.mass,
                radius=item.radius,
                owner_id=item.owner_id,
                ttl=item.ttl,
            )
            for item in self.world.ejected.values()
        )
        viruses = tuple(
            VirusView(id=v.id, x=v.x, y=v.y, mass=v.mass, radius=v.radius)
            for v in self.world.viruses.values()
        )

        return (players_by_id, tuple(players), foods, ejected, viruses)

    def _fallback_action(self, me: PlayerView) -> BotAction:
        if me.blobs:
            first = me.blobs[0]
            return BotAction(target_x=first.x, target_y=first.y)
        return BotAction(target_x=config.WORLD_WIDTH / 2.0, target_y=config.WORLD_HEIGHT / 2.0)

    def _build_name(self, spec: BotSpec, index: int) -> str:
        prefix = spec.name_prefix or spec.plugin_name.replace("_", " ").title()
        return f"{prefix}-{index + 1}"

    def _spawn_bot(
        self,
        *,
        plugin_name: str,
        team_id: str | None,
        name_prefix: str,
        now: float,
        explicit_name: str | None = None,
    ) -> str:
        self._spawn_index += 1
        sequence = self._spawn_index
        bot_name = explicit_name or f"{name_prefix}-{sequence}"
        player = self.world.add_player(
            player_name=bot_name,
            now=now,
            is_bot=True,
            bot_plugin=plugin_name,
            bot_team=team_id,
        )
        try:
            init_ctx = BotInitContext(
                plugin_name=plugin_name,
                bot_name=player.name,
                team_id=team_id,
                bot_index=sequence,
                rng=random.Random(self._rng.randint(0, 2_000_000_000)),
            )
            brain = self._registry.create(plugin_name, init_ctx)
        except Exception:
            self.world.remove_player(player.id)
            raise

        self._agents[player.id] = _BotAgent(
            player_id=player.id,
            plugin_name=plugin_name,
            team_id=team_id,
            name_prefix=name_prefix,
            brain=brain,
            memory={"_was_alive": True},
        )
        return player.id

    def _spawn_extra_on_elimination(self, eliminated: _BotAgent, now: float) -> None:
        if not config.BOT_SPAWN_ON_EATEN:
            return
        if config.BOT_SPAWN_PER_ELIMINATION <= 0:
            return

        max_active = max(1, config.BOT_MAX_ACTIVE)
        spawned = 0
        for _ in range(config.BOT_SPAWN_PER_ELIMINATION):
            if len(self._agents) >= max_active:
                break
            try:
                self._spawn_bot(
                    plugin_name=eliminated.plugin_name,
                    team_id=eliminated.team_id,
                    name_prefix=eliminated.name_prefix,
                    now=now,
                )
                spawned += 1
            except Exception:
                logger.exception(
                    "Failed to spawn extra bot for eliminated player %s using plugin '%s'",
                    eliminated.player_id,
                    eliminated.plugin_name,
                )
                break

        if spawned > 0:
            logger.info(
                "Bot '%s' was eliminated. Spawned %d extra bot(s). Active bots: %d/%d",
                eliminated.player_id,
                spawned,
                len(self._agents),
                max_active,
            )

    def _team_key(self, plugin_name: str, team_id: str | None) -> str:
        return f"{plugin_name}:{team_id or '-'}"

    @staticmethod
    def _clamp(value: float, min_value: float, max_value: float) -> float:
        return min(max(value, min_value), max_value)

File: cpp_backend/src/bots/manager.cpp
#include "manager.hpp"

#include <algorithm>
#include <cctype>
#include <sstream>
#include <stdexcept>
#include <thread>

namespace agario::bots {
namespace {

std::string trim_copy(const std::string& input) {
  std::size_t start = input.find_first_not_of(" \t\n\r");
  if (start == std::string::npos) {
    return "";
  }
  std::size_t end = input.find_last_not_of(" \t\n\r");
  return input.substr(start, end - start + 1);
}

std::string lower_copy(const std::string& input) {
  std::string out = input;
  for (char& ch : out) {
    ch = static_cast<char>(std::tolower(static_cast<unsigned char>(ch)));
  }
  return out;
}

std::string title_case(const std::string& input) {
  std::string out;
  out.reserve(input.size());
  bool start_word = true;
  for (char ch : input) {
    if (ch == '_') {
      out.push_back(' ');
      start_word = true;
      continue;
    }
    if (start_word) {
      out.push_back(static_cast<char>(std::toupper(static_cast<unsigned char>(ch))));
      start_word = false;
    } else {
      out.push_back(static_cast<char>(std::tolower(static_cast<unsigned char>(ch))));
    }
  }
  return out;
}

}  // namespace

std::vector<BotSpec> parse_bot_specs(const std::string& raw) {
  std::vector<BotSpec> specs;
  std::string cleaned = trim_copy(raw);
  if (cleaned.empty()) {
    return specs;
  }

  std::stringstream list_stream(cleaned);
  std::string chunk;
  while (std::getline(list_stream, chunk, ',')) {
    std::string part = trim_copy(chunk);
    if (part.empty()) {
      continue;
    }
    std::stringstream spec_stream(part);
    std::string token;
    std::vector<std::string> tokens;
    while (std::getline(spec_stream, token, ':')) {
      tokens.push_back(trim_copy(token));
    }
    if (tokens.empty() || tokens[0].empty()) {
      throw std::invalid_argument("Invalid bot spec '" + part + "': plugin name is required");
    }
    BotSpec spec;
    spec.plugin_name = lower_copy(tokens[0]);
    spec.count = 1;
    if (tokens.size() >= 2 && !tokens[1].empty()) {
      spec.count = std::stoi(tokens[1]);
    }
    if (tokens.size() >= 3 && !tokens[2].empty() && tokens[2] != "-") {
      spec.team_id = tokens[2];
    }
    if (tokens.size() >= 4 && !tokens[3].empty() && tokens[3] != "-") {
      spec.name_prefix = tokens[3];
    }
    if (spec.count <= 0) {
      throw std::invalid_argument("Invalid bot spec '" + part + "': count must be > 0");
    }
    specs.push_back(std::move(spec));
  }
  return specs;
}

BotManager::BotManager(GameWorld& world,
                       bool enabled,
                       std::vector<std::string> plugin_modules,
                       std::vector<BotSpec> bot_specs,
                       int seed,
                       int bot_threads)
    : world_(world),
      enabled_(enabled),
      plugin_modules_(std::move(plugin_modules)),
      bot_specs_(std::move(bot_specs)),
      rng_(static_cast<std::mt19937::result_type>(seed)),
      bot_threads_(bot_threads) {}

BotManager BotManager::from_config(GameWorld& world) {
  const auto& cfg = config::get();
  return BotManager(
      world,
      cfg.bots_enabled,
      cfg.bot_plugin_modules,
      parse_bot_specs(cfg.bot_specs),
      cfg.bot_random_seed,
      cfg.bot_threads);
}

boost::json::object BotManager::describe() const {
  const auto& cfg = config::get();
  boost::json::object payload;
  payload["enabled"] = enabled_;
  payload["started"] = started_;

  boost::json::array plugin_modules;
  for (const auto& mod : plugin_modules_) {
    plugin_modules.push_back(boost::json::value(mod));
  }
  payload["pluginModules"] = std::move(plugin_modules);

  boost::json::array registered;
  for (const auto& name : registry_.names()) {
    registered.push_back(boost::json::value(name));
  }
  payload["registeredPlugins"] = std::move(registered);

  boost::json::array specs;
  for (const auto& spec : bot_specs_) {
    boost::json::object row;
    row["plugin"] = spec.plugin_name;
    row["count"] = spec.count;
    row["team"] = spec.team_id.empty() ? boost::json::value(nullptr) : boost::json::value(spec.team_id);
    row["namePrefix"] = spec.name_prefix.empty() ? boost::json::value(nullptr) : boost::json::value(spec.name_prefix);
    specs.push_back(std::move(row));
  }
  payload["botSpecs"] = std::move(specs);
  payload["activeBots"] = static_cast<int>(agents_.size());
  payload["spawnOnEaten"] = cfg.bot_spawn_on_eaten;
  payload["spawnPerElimination"] = cfg.bot_spawn_per_elimination;
  payload["maxActiveBots"] = cfg.bot_max_active;
  return payload;
}

void BotManager::ensure_started(double now) {
  if (started_ || !enabled_) {
    return;
  }
  register_core_plugins(registry_);
  for (const auto& spec : bot_specs_) {
    for (int i = 0; i < spec.count; ++i) {
      std::string name_prefix = !spec.name_prefix.empty() ? spec.name_prefix : title_case(spec.plugin_name);
      std::string player_name = build_name(spec, i);
      spawn_bot(spec.plugin_name, spec.team_id, name_prefix, now, player_name);
    }
  }
  started_ = true;
}

void BotManager::tick(double dt, double now) {
  if (!started_ || agents_.empty()) {
    return;
  }

  auto views = build_views();
  if (views.players_by_id.empty()) {
    return;
  }

  std::vector<SpawnRequest> pending_spawns;

  std::unordered_map<std::string, std::vector<std::string>> team_members;
  for (const auto& entry : agents_) {
    const auto& agent = entry.second;
    std::string key = team_key(agent.plugin_name, agent.team_id, agent.player_id);
    team_members[key].push_back(agent.player_id);
  }
  for (auto& entry : team_members) {
    auto& members = entry.second;
    members.erase(std::remove_if(members.begin(), members.end(), [&](const std::string& pid) {
      return views.players_by_id.find(pid) == views.players_by_id.end();
    }), members.end());
    std::sort(members.begin(), members.end());
    auto& team_ptr = team_state_[entry.first];
    if (!team_ptr) {
      team_ptr = std::make_shared<TeamData>();
    }
    auto& team = *team_ptr;
    std::lock_guard<std::mutex> guard(team.mutex);
    team.state.set("members", members);
  }

  std::vector<BotAgent*> active_agents;
  std::vector<PlayerView> active_views;
  std::vector<TeamData*> team_refs;
  active_agents.reserve(agents_.size());
  active_views.reserve(agents_.size());
  team_refs.reserve(agents_.size());

  for (auto it = agents_.begin(); it != agents_.end();) {
    auto& agent = it->second;
    auto player_it = views.players_by_id.find(agent.player_id);
    if (player_it == views.players_by_id.end()) {
      it = agents_.erase(it);
      continue;
    }
    const PlayerView& me = player_it->second;

    bool was_alive = agent.memory.get_or<bool>("_was_alive", !me.blobs.empty());
    bool is_alive = !me.blobs.empty();
    if (!is_alive && was_alive) {
      pending_spawns.push_back({agent.plugin_name, agent.team_id, agent.name_prefix});
    }
    agent.memory.set("_was_alive", is_alive);

    active_agents.push_back(&agent);
    active_views.push_back(me);
    std::string key = team_key(agent.plugin_name, agent.team_id, agent.player_id);
    auto team_it = team_state_.find(key);
    if (team_it == team_state_.end() || !team_it->second) {
      auto& team_ptr = team_state_[key];
      if (!team_ptr) {
        team_ptr = std::make_shared<TeamData>();
      }
      team_refs.push_back(team_ptr.get());
    } else {
      team_refs.push_back(team_it->second.get());
    }
    ++it;
  }

  struct ActionResult {
    std::string player_id;
    BotAction action;
  };

  std::vector<ActionResult> results(active_agents.size());
  int thread_count = bot_threads_ > 0 ? bot_threads_ : static_cast<int>(std::thread::hardware_concurrency());
  if (thread_count <= 0) {
    thread_count = 1;
  }
  if (thread_count > static_cast<int>(active_agents.size())) {
    thread_count = static_cast<int>(active_agents.size());
  }
  if (thread_count <= 1) {
    thread_count = 1;
  }

  auto worker = [&](std::size_t start, std::size_t end) {
    for (std::size_t idx = start; idx < end; ++idx) {
      BotAgent* agent = active_agents[idx];
      const PlayerView& me = active_views[idx];
      TeamData* team = team_refs[idx];
      if (!team) {
        results[idx] = {agent->player_id, fallback_action(me)};
        continue;
      }

      BotContext ctx;
      ctx.now = now;
      ctx.dt = dt;
      ctx.world_width = config::get().world_width;
      ctx.world_height = config::get().world_height;
      ctx.me = me;
      ctx.players = views.players;
      ctx.foods = views.foods;
      ctx.ejected = views.ejected;
      ctx.viruses = views.viruses;
      ctx.team_state = &team->state;
      ctx.memory = &agent->memory;

      BotAction action;
      try {
        std::lock_guard<std::mutex> guard(team->mutex);
        action = agent->brain->decide(ctx);
      } catch (...) {
        action = fallback_action(me);
      }

      results[idx] = {agent->player_id, action};
    }
  };

  if (thread_count == 1) {
    worker(0, active_agents.size());
  } else {
    std::vector<std::thread> threads;
    threads.reserve(static_cast<std::size_t>(thread_count));
    std::size_t total = active_agents.size();
    std::size_t chunk = (total + static_cast<std::size_t>(thread_count) - 1) / static_cast<std::size_t>(thread_count);
    std::size_t start = 0;
    for (int i = 0; i < thread_count; ++i) {
      std::size_t end = std::min(total, start + chunk);
      if (start >= end) {
        break;
      }
      threads.emplace_back(worker, start, end);
      start = end;
    }
    for (auto& t : threads) {
      t.join();
    }
  }

  for (const auto& result : results) {
    world_.set_input(
        result.player_id,
        clamp(result.action.target_x, 0.0, config::get().world_width),
        clamp(result.action.target_y, 0.0, config::get().world_height),
        result.action.split,
        result.action.eject);
  }

  for (const auto& request : pending_spawns) {
    spawn_extra_on_elimination(request, now);
  }
}

BotManager::ViewPack BotManager::build_views() const {
  ViewPack pack;
  for (const auto& pkv : world_.players) {
    const Player& player = pkv.second;
    PlayerView view;
    view.id = player.id;
    view.name = player.name;
    view.color = player.color;
    view.is_bot = player.is_bot;
    view.plugin_name = player.bot_plugin;
    view.team_id = player.bot_team;
    view.total_mass = 0.0;
    for (const auto& bkv : player.blobs) {
      const Blob& blob = bkv.second;
      BlobView b;
      b.id = blob.id;
      b.player_id = blob.player_id;
      b.x = blob.x;
      b.y = blob.y;
      b.mass = blob.mass;
      b.radius = blob.radius();
      view.total_mass += b.mass;
      view.blobs.push_back(std::move(b));
    }
    pack.players.push_back(std::move(view));
  }
  std::sort(pack.players.begin(), pack.players.end(), [](const PlayerView& a, const PlayerView& b) {
    return a.id < b.id;
  });
  for (const auto& player : pack.players) {
    pack.players_by_id[player.id] = player;
  }

  for (const auto& fkv : world_.foods) {
    const Food& food = fkv.second;
    FoodView fv;
    fv.id = food.id;
    fv.x = food.x;
    fv.y = food.y;
    fv.mass = food.mass;
    fv.radius = food.radius();
    fv.color = food.color;
    pack.foods.push_back(std::move(fv));
  }

  for (const auto& ekv : world_.ejected) {
    const EjectedMass& item = ekv.second;
    EjectedView ev;
    ev.id = item.id;
    ev.x = item.x;
    ev.y = item.y;
    ev.mass = item.mass;
    ev.radius = item.radius();
    ev.owner_id = item.owner_id;
    ev.ttl = item.ttl;
    pack.ejected.push_back(std::move(ev));
  }

  for (const auto& vkv : world_.viruses) {
    const Virus& virus = vkv.second;
    VirusView vv;
    vv.id = virus.id;
    vv.x = virus.x;
    vv.y = virus.y;
    vv.mass = virus.mass;
    vv.radius = virus.radius();
    pack.viruses.push_back(std::move(vv));
  }

  return pack;
}

BotAction BotManager::fallback_action(const PlayerView& me) const {
  if (!me.blobs.empty()) {
    const auto& first = me.blobs.front();
    return {first.x, first.y, false, false};
  }
  return {config::get().world_width * 0.5, config::get().world_height * 0.5, false, false};
}

std::string BotManager::build_name(const BotSpec& spec, int index) const {
  std::string prefix = !spec.name_prefix.empty() ? spec.name_prefix : title_case(spec.plugin_name);
  return prefix + "-" + std::to_string(index + 1);
}

std::string BotManager::spawn_bot(const std::string& plugin_name,
                                  const std::string& team_id,
                                  const std::string& name_prefix,
                                  double now,
                                  const std::string& explicit_name) {
  spawn_index_ += 1;
  int sequence = spawn_index_;
  std::string bot_name = !explicit_name.empty() ? explicit_name : name_prefix + "-" + std::to_string(sequence);
  Player& player = world_.add_player(bot_name, now, true, plugin_name, team_id);

  std::uniform_int_distribution<int> seed_dist(0, 2'000'000'000);
  BotInitContext init_ctx;
  init_ctx.plugin_name = plugin_name;
  init_ctx.bot_name = player.name;
  init_ctx.team_id = team_id;
  init_ctx.bot_index = sequence;
  init_ctx.rng = std::mt19937(static_cast<std::mt19937::result_type>(seed_dist(rng_)));

  std::unique_ptr<BotBrain> brain;
  try {
    brain = registry_.create(plugin_name, init_ctx);
  } catch (...) {
    world_.remove_player(player.id);
    throw;
  }
  BotAgent agent;
  agent.player_id = player.id;
  agent.plugin_name = plugin_name;
  agent.team_id = team_id;
  agent.name_prefix = name_prefix;
  agent.brain = std::move(brain);
  agent.memory.set("_was_alive", true);
  agents_[player.id] = std::move(agent);
  return player.id;
}

void BotManager::spawn_extra_on_elimination(const SpawnRequest& eliminated, double now) {
  const auto& cfg = config::get();
  if (!cfg.bot_spawn_on_eaten || cfg.bot_spawn_per_elimination <= 0) {
    return;
  }
  int max_active = std::max(1, cfg.bot_max_active);
  int spawned = 0;
  for (int i = 0; i < cfg.bot_spawn_per_elimination; ++i) {
    if (static_cast<int>(agents_.size()) >= max_active) {
      break;
    }
    try {
      spawn_bot(eliminated.plugin_name, eliminated.team_id, eliminated.name_prefix, now, "");
      spawned += 1;
    } catch (...) {
      break;
    }
  }
  (void)spawned;
}

std::string BotManager::team_key(const std::string& plugin_name, const std::string& team_id, const std::string& player_id) const {
  if (team_id.empty()) {
    return plugin_name + ":" + player_id;
  }
  return plugin_name + ":" + team_id;
}

double BotManager::clamp(double value, double min_value, double max_value) {
  return std::min(std::max(value, min_value), max_value);
}

}  // namespace agario::bots
